### DISCUSSION TITLE
"toomanyrequests: retry-after: 3.205835ms, allowed: 2000/minute" error while pushing  docker images to GHCR

### DISCUSSION THREAD

**Discussion by marcogiorgio** [community/community#42479] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi, 
I have a CI running on GitHub actions that tags and pushes PHP images to GHCR. The part of the CI that does this is configured as a matrix job which currently generates 28 jobs (for different php/alpine versions).
Each one of these jobs executes the command `docker push ghcr.io/USERNAME/NAME/TAG`. 
Since I have multiple tags, each job executes the command multiple times. 
The problem is that the workflow keeps failing on almost every matrix job with the error `toomanyrequests: retry-after: Xms, allowed: 2000/minute` on the `docker push` command. I am currently authenticating to GHCR using `${{ github.actor }}` and `${{ secrets.GITHUB_TOKEN }}`. 
I can't find anywhere this kind of limitation and I don't know if I can raise it and where. Do you have any suggestions?
Thanks

**Comment by DSDmark** [comment, marked as answer]
Hi @marcogiorgio,

I think, the problem you are facing is **rate limiting** issue when pushing multiple **Docker** images to GitHub Container Registry `(GHCR)`.

> `GHCR` has a rate limit of `2,000` requests per minute for **authenticated requests**, and it is possible that your CI workflow is exceeding this limit.

You could try these steps:-

  -  Reduce the number of images being pushed in each job by limiting the matrix to a smaller set of `PHP/Alpine` versions.
  -  Use the `--max-concurrent-uploads` flag when running the docker push command to limit the number of concurrent pushes.
  -  Use a queue or other mechanism to spread out the pushes over time, rather than trying to push all the images at once.
  -  If you are using multiple tokens or service accounts to authenticate the pushes, consider using a single token or service account to consolidate the pushes and avoid hitting the rate limit.

**Reply by marcogiorgio** [reply]
Hi!
Thanks for your quick reply! May you link where you found that line of docs? I was not able to find it. I tried the `--max-concurrent-uploads` and it seems to work actually so I'll probably use that! Thanks again and happy holiday!
